x/posts/keeper/polls.go

```diff
@@ -0,0 +1,114 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/desmos-labs/desmos/v3/x/posts/types"
+)
+
+// HasPoll tells whether the specified post contains a poll with the provided id
+func (k Keeper) HasPoll(ctx sdk.Context, subspaceID uint64, postID uint64, pollID uint32) bool {
+	attachment, found := k.GetAttachment(ctx, subspaceID, postID, pollID)
+	if !found {
+		return false
+	}
+	return types.IsPoll(attachment)
+}
+
+// GetPoll returns the poll having the given id.
+// If not poll with the given id is found, the function returns nil and false.
+func (k Keeper) GetPoll(ctx sdk.Context, subspaceID uint64, postID uint64, pollID uint32) (poll *types.Poll, found bool) {
+	attachment, found := k.GetAttachment(ctx, subspaceID, postID, pollID)
+	if !found {
+		return nil, false
+	}
+
+	poll, ok := attachment.Content.GetCachedValue().(*types.Poll)
+	return poll, ok
+}
+
+// Tally iterates over the votes and returns the tally results of a poll
+func (k Keeper) Tally(ctx sdk.Context, subspaceID uint64, postID uint64, pollID uint32) *types.PollTallyResults {
+	poll, found := k.GetPoll(ctx, subspaceID, postID, pollID)
+	if !found {
+		return nil
+	}
+
+	// Create the map index -> count(votes)
+	results := make(map[uint32]uint64, len(poll.ProvidedAnswers))
+	for i := range poll.ProvidedAnswers {
+		results[uint32(i)] = 0
+	}
+
+	k.IteratePollUserAnswers(ctx, subspaceID, postID, pollID, func(_ int64, answer types.UserAnswer) (stop bool) {
+		// Update the results
+		for _, answerIndex := range answer.AnswersIndexes {
+			results[answerIndex]++
+		}
+
+		// Delete the user answer
+		k.DeleteUserAnswer(ctx, answer.SubspaceID, answer.PostID, answer.PollID, answer.User)
```

Comment:
The `PollAnswers` query will always return empty array if we remove it after tally. Should we remain `UserAnswer`.

Reply:
This is done on purpose in order to avoid filling the store with useless data that can be queried elsewhere. Even after deleted, clients will be able to get the user answers by querying for the following event: 
```
answer_poll.subspace_id={subspace_id}&answer_poll.post_id={post_id}&answer_poll.poll_id={poll_id}
```

This is the same reason why the Cosmos SDK removes a proposal's votes after the results have been tallied